napari/_vispy/vispy_transforms.py

```diff
@@ -0,0 +1,82 @@
+import numpy as np
+from vispy.visuals.transforms import STTransform, ChainTransform
+
+
+class VispyTransformChain:
+    """Class containing an ordered sequence of vispy transforms.
+
+    Parameters
+    ----------
+    napari_transform_chain : napari.layers.transforms.TransformChain
+        Transform chain for an ordered sequence of napari transforms.
+
+    Attributes
+    ----------
+    scale : ndarray, shape(4)
+        Combined scale of all chained transforms, vispy format (X,Y,Z,T).
+    simplified_transform : vispy.visuals.transforms.ChainTransform
+        Simplified ChainTransform where transforms have been combined into one.
+        Used to reduce compute time, since a single matrix operation is faster.
+    transform_chain : vispy.visuals.transforms.ChainTransform
+        Transform chain for an ordered sequence of vispy transforms.
+    translate : ndarray, shape(4)
+        Combined tranlation of all chained transforms, vispy format (X,Y,Z,T).
+    """
+
+    def __init__(self, napari_transform_chain):
+        self.napari_transform_chain = napari_transform_chain
+
+        self.transform_chain = self._vispy_transform_chain()
+        self._calculate_attributes()
+
+        self.napari_transform_chain.events.added.connect(self._add)
+        self.napari_transform_chain.events.removed.connect(self._remove)
+        self.transform_chain.events.reordered.connect(self._reorder)
+
+    def _vispy_transform_chain(self):
+        """Builds vispy ChainTransform from a napari TransformChain."""
+        vispy_transforms = []
+        for t in self.napari_transform_chain:
+            # From napari format (t, z, y, x)
+            # to the vispy format (x, y, z, ?)
+            translate = np.flip(t.translate)  # vispy ordered array
+            scale = np.flip(t.scale)  # vispy ordered array
```

Comment:
Is my understanding here correct? I think the napari transforms have the format (T,Z,Y,X), and the vispy transforms have the format (X,Y,Z,T). Additionally napari transform scale and translate arrays can have fewer than 4 elements, eg if you're just specifying a translation in x and y.